File: fons/codegen/copia.ts
/**
 * Copia Method Registry - Unified translations for Latin set methods
 *
 * COMPILER PHASE
 * ==============
 * codegen (all targets)
 *
 * ARCHITECTURE
 * ============
 * This module defines translations for copia<T> (HashSet) methods across all targets.
 * Phase 2: Zig only. Other targets added in Phase 3.
 *
 * WHY UNIFIED REGISTRY
 * ====================
 * - Single source of truth for method behavior across targets
 * - needsAlloc flag determines if codegen passes curator (allocator)
 * - Reduces N methods x M targets file sprawl
 *
 * ZIG SPECIFICS
 * =============
 * Zig has no native Set type. We implement copia as HashMap(T, void).
 * - copia<textus> maps to CopiaTextus (StringHashMap wrapper)
 * - copia<T> maps to CopiaAuto(T) (AutoHashMap wrapper)
 */

// =============================================================================
// TYPES
// =============================================================================

/**
 * Generator function type for Zig collection methods.
 * WHY: curator parameter is the current allocator from curatorStack.
 */
export type ZigGenerator = (obj: string, args: string[], curator: string) => string;

/**
 * Describes how to translate a Latin method.
 */
export interface CopiaMethod {
    /** True if method mutates the collection in place */
    mutates: boolean;

    /** True if method needs allocator (growing, returning new collection) */
    needsAlloc: boolean;

    /**
     * Zig translation.
     * - string: method name to delegate to stdlib
     * - function: custom code generation
     */
    zig?: string | ZigGenerator;

    // Future phases will add: ts, py, rs, cpp
}

// =============================================================================
// METHOD REGISTRY
// =============================================================================

/**
 * Registry of Latin set methods with translations.
 *
 * Allocator categories:
 * - Adding (adde): Yes - may need to resize
 * - Reading (habet, longitudo, vacua): No
 * - Removing (dele, purga): No - doesn't allocate
 * - Iteration (valores): No
 * - Set operations (unio, intersectio, etc.): Yes - return new sets
 */
export const COPIA_METHODS: Record<string, CopiaMethod> = {
    // -------------------------------------------------------------------------
    // CORE OPERATIONS
    // -------------------------------------------------------------------------

    /** Add element (mutates, needs allocator for growth) */
    adde: {
        mutates: true,
        needsAlloc: true,
        zig: (obj, args, curator) => `${obj}.adde(${curator}, ${args[0]})`,
    },

    /** Check if element exists */
    habet: {
        mutates: false,
        needsAlloc: false,
        zig: (obj, args) => `${obj}.habet(${args[0]})`,
    },

    /** Delete element (mutates) */
    dele: {
        mutates: true,
        needsAlloc: false,
        zig: (obj, args) => `_ = ${obj}.dele(${args[0]})`,
    },

    /** Get size */
    longitudo: {
        mutates: false,
        needsAlloc: false,
        zig: obj => `${obj}.longitudo()`,
    },

    /** Check if empty */
    vacua: {
        mutates: false,
        needsAlloc: false,
        zig: obj => `${obj}.vacua()`,
    },

    /** Clear all elements (mutates) */
    purga: {
        mutates: true,
        needsAlloc: false,
        zig: obj => `${obj}.purga()`,
    },

    // -------------------------------------------------------------------------
    // ITERATION
    // -------------------------------------------------------------------------

    /** Get values iterator (for sets, this iterates elements) */
    valores: {
        mutates: false,
        needsAlloc: false,
        zig: obj => `${obj}.valores()`,
    },

    /** Iterate with callback */
    perambula: {
        mutates: false,
        needsAlloc: false,
        zig: () => `@compileError("perambula not implemented for Zig - use 'ex set.valores() pro item { ... }' loop")`,
    },

    // -------------------------------------------------------------------------
    // SET OPERATIONS - NOT IMPLEMENTED
    // These require creating new sets with allocator, complex to generate inline
    // -------------------------------------------------------------------------

    /** Union: A U B */
    unio: {
        mutates: false,
        needsAlloc: true,
        zig: () => `@compileError("unio not implemented for Zig - use explicit loop to merge sets")`,
    },

    /** Intersection: A n B */
    intersectio: {
        mutates: false,
        needsAlloc: true,
        zig: () => `@compileError("intersectio not implemented for Zig - use explicit loop")`,
    },

    /** Difference: A - B */
    differentia: {
        mutates: false,
        needsAlloc: true,
        zig: () => `@compileError("differentia not implemented for Zig - use explicit loop")`,
    },

    /** Symmetric difference */
    symmetrica: {
        mutates: false,
        needsAlloc: true,
        zig: () => `@compileError("symmetrica not implemented for Zig - use explicit loop")`,
    },

    // -------------------------------------------------------------------------
    // PREDICATES - NOT IMPLEMENTED
    // -------------------------------------------------------------------------

    /** Is subset of other */
    subcopia: {
        mutates: false,
        needsAlloc: false,
        zig: () => `@compileError("subcopia not implemented for Zig - use explicit loop")`,
    },

    /** Is superset of other */
    supercopia: {
        mutates: false,
        needsAlloc: false,
        zig: () => `@compileError("supercopia not implemented for Zig - use explicit loop")`,
    },

    // -------------------------------------------------------------------------
    // CONVERSIONS - NOT IMPLEMENTED
    // -------------------------------------------------------------------------

    /** Convert to list */
    inLista: {
        mutates: false,
        needsAlloc: true,
        zig: () => `@compileError("inLista not implemented for Zig - iterate with ex...pro into ArrayList")`,
    },
};

// =============================================================================
// LOOKUP FUNCTIONS
// =============================================================================

/**
 * Look up a Latin method name and return its definition.
 */
export function getCopiaMethod(name: string): CopiaMethod | undefined {
    return COPIA_METHODS[name];
}

/**
 * Check if a method name is a known copia method.
 */
export function isCopiaMethod(name: string): boolean {
    return name in COPIA_METHODS;
}

File: subsidia/zig/copia.zig
// Copia - HashSet wrapper for Faber
// Wraps std.AutoHashMap(T, void) as a set type
//
// WHY: No stored allocator. Methods that need allocation accept it as parameter.
// This ensures cura...fit blocks properly thread the current allocator.
//
// WHY: Zig has no native Set type. We use HashMap with void values.
// For string elements, use CopiaTextus which wraps StringHashMap(void).

const std = @import("std");

# Copia(T) is a hash set with Latin-named methods.
# For string elements, use CopiaTextus which wraps StringHashMap(void).
# For other types, use CopiaAuto(T) which wraps AutoHashMap(T, void).
pub fn Copia(comptime T: type) type {
    if (T == []const u8) {
        return CopiaTextus;
    }
    return CopiaAuto(T);
}

# Copia for string elements - wraps std.StringHashMap(void)
pub const CopiaTextus = struct {
    map: std.StringHashMap(void),

    const Self = @This();

    // =====================================================================
    // CONSTRUCTION
    // =====================================================================

    # Create an empty copia.
    pub fn init(alloc: std.mem.Allocator) Self {
        return .{ .map = std.StringHashMap(void).init(alloc) };
    }

    # Free the copia's memory.
    pub fn deinit(self: *Self) void {
        self.map.deinit();
    }

    // =====================================================================
    // CORE OPERATIONS
    // =====================================================================

    # Add element (mutates, needs allocator for growth).
    pub fn adde(self: *Self, alloc: std.mem.Allocator, value: []const u8) void {
        _ = alloc; // StringHashMap tracks its allocator internally
        self.map.put(value, {}) catch @panic("OOM");
    }

    # Check if element exists.
    pub fn habet(self: Self, value: []const u8) bool {
        return self.map.contains(value);
    }

    # Delete element (mutates).
    pub fn dele(self: *Self, value: []const u8) bool {
        return self.map.remove(value);
    }

    # Get size.
    pub fn longitudo(self: Self) usize {
        return self.map.count();
    }

    # Check if empty.
    pub fn vacua(self: Self) bool {
        return self.map.count() == 0;
    }

    # Clear all elements (mutates).
    pub fn purga(self: *Self) void {
        self.map.clearRetainingCapacity();
    }

    // =====================================================================
    // ITERATION
    // =====================================================================

    # Get values (keys) iterator.
    pub fn valores(self: *Self) std.StringHashMap(void).KeyIterator {
        return self.map.keyIterator();
    }
};

# Copia for non-string elements - wraps std.AutoHashMap(T, void)
pub fn CopiaAuto(comptime T: type) type {
    return struct {
        map: std.AutoHashMap(T, void),

        const Self = @This();

        // =====================================================================
        // CONSTRUCTION
        // =====================================================================

        # Create an empty copia.
        pub fn init(alloc: std.mem.Allocator) Self {
            return .{ .map = std.AutoHashMap(T, void).init(alloc) };
        }

        # Free the copia's memory.
        pub fn deinit(self: *Self) void {
            self.map.deinit();
        }

        // =====================================================================
        // CORE OPERATIONS
        // =====================================================================

        # Add element (mutates, needs allocator for growth).
        pub fn adde(self: *Self, alloc: std.mem.Allocator, value: T) void {
            _ = alloc; // AutoHashMap tracks its allocator internally
            self.map.put(value, {}) catch @panic("OOM");
        }

        # Check if element exists.
        pub fn habet(self: Self, value: T) bool {
            return self.map.contains(value);
        }

        # Delete element (mutates).
        pub fn dele(self: *Self, value: T) bool {
            return self.map.remove(value);
        }

        # Get size.
        pub fn longitudo(self: Self) usize {
            return self.map.count();
        }

        # Check if empty.
        pub fn vacua(self: Self) bool {
            return self.map.count() == 0;
        }

        # Clear all elements (mutates).
        pub fn purga(self: *Self) void {
            self.map.clearRetainingCapacity();
        }

        // =====================================================================
        // ITERATION
        // =====================================================================

        # Get values (keys) iterator.
        pub fn valores(self: *Self) std.AutoHashMap(T, void).KeyIterator {
            return self.map.keyIterator();
        }
    };
}
